src/interval.jl

```diff
@@ -71,7 +71,27 @@ struct Interval{T} <: AbstractInterval{T}
     end
 end
 
-Interval{T}(f, l, inc::Inclusivity) where T = Interval{T}(convert(T, f), convert(T, l), inc)
+isbounded(a) = !isposinf(a) && !isneginf(a)
+isunbounded(a) = !isbounded(a)
+
+function Interval{T}(f, l, inc::Inclusivity) where T
+    if (isbounded(f) && isbounded(l)) || (isunbounded(f) && isunbounded(l))
+        return Interval{T}(convert(T, f), convert(T, l), inc)
+    else
+        # If either endpoint is unbounded, we want to convert the bounded variable, and then
+        # try and promote them both to a compatable type.
+        # If T is a subset of the Infinite type, then don't try to convert at all, as trying
+        # to convert any type to Infinite will result in an error
+        if !(T <: Infinite)
+            if isbounded(f)
+                f = convert(T, f)
+            else
+                l = convert(T, l)
+            end
+        end
+        return Interval(f, l, inc)
+    end
+end
```

Comment:
I think if a user has explicitly set `T` we shouldn't be messing with it.

Reply:
hm yeah I suppose this change is actually just a byproduct of trying to throw in unbounded values into the tests.

This test is specifically the issue (https://github.com/invenia/Intervals.jl/blob/fc/using-infinity/test/interval.jl#L46-L47):
```
@test Interval(b, a, Inclusivity(true, false)) == Interval{typeof(a)}(a, b, Inclusivity(false, true))
```
When you create the interval on the left size of `==` it will `promote(b, a)` so if one value is `∞` and the other is a value, lets say `10`, then it converts both values to `InfExtended`. Now on the right side of `==`, if a is of type `Infinite` (due to being a `∞` or `-∞`) then it will try to `convert(Infinite, 10)` which is an error. On the other side of the spectrum, if a is `10` which would make it `Int64` type, then trying to do `convert(Int64, ∞)` will also fail. I guess we can just let it be an error if somebody puts in the wrong type in the Interval constructor.

Reply:
I think the test should be updated to be using:
```julia
julia> promote_type(Infinite, Int)
InfExtended{Int64}
```
Then in the tests you can use `Interval{promote_type(typeof(a), typeof(b))}(...)`